File: client/src/models/AIFunction.ts
import { z } from "zod"
import InputVariable from "./InputVariable"
import OutputAssertion from "./OutputAssertion"

const AIFunctionRouteInput = z.object({
  name: z.string().min(1).max(40),
  description: z.string().min(1).max(1000),
  input_variables: z.array(InputVariable),
  output_assertions: z.array(OutputAssertion),
  dataset: z.array(z.record(z.string())),
})
const AIFunction = AIFunctionRouteInput.extend({
  number_of_prompts: z.number().nonnegative(),
  username: z.string().email(),
  creation_time: z.string().refine(
    (val) => {
      const date = new Date(val)
      return !isNaN(date.getTime())
    },
    {
      message: "Invalid datetime format",
    }
  ),
})

export { AIFunctionRouteInput, AIFunction }

File: backend/App/models/ai_function.py
from pydantic import (
    BaseModel,
    StringConstraints,
    Field,
    NonNegativeInt,
    EmailStr,
)
from typing import Annotated, List, Dict
from .objectID import PydanticObjectId
from datetime import datetime
from .output_assertion import OutputAssertion


class InputVariable(BaseModel):
    name: Annotated[str, StringConstraints(min_length=1, max_length=40)]


class AIFunctionRouteInput(BaseModel):
    name: Annotated[str, StringConstraints(min_length=1, max_length=40)] = Field(
        ..., example="Summarize Texts"
    )

    description: Annotated[str, StringConstraints(min_length=1, max_length=1000)] = (
        Field(..., example="Summarizes english texts to a given number of sentences.")
    )

    input_variables: List[InputVariable]

    output_assertions: List[OutputAssertion]

    dataset: List[Dict[str, str]]


class AIFunction(AIFunctionRouteInput):
    number_of_prompts: NonNegativeInt
    username: EmailStr
    creation_time: datetime


class AIFunctionWithID(AIFunction):
    id: PydanticObjectId = Field(alias="_id")


class AIFunctionList(BaseModel):
    ai_function_list: List[AIFunctionWithID]
